README.md
Herramientas Churreras
======================

### (c) 2013 - 2017 Alfonso Saavedra "Son Link"
##### Churrera (c) 2013 [The Mojon twins](http://www.mojontwins.com/)

Scripts bajo licencia GPLv3. Ver COPYING para mas información.

##### Churrera Installer
Instalador para las herramientas necesarias para la Churrera de los Mojon Twins.

##### Churrera Retriever
Crea o actualiza un proyecto de la churrera

##### make.sh
Script para compilar el juego.

La Churrera te permita hacer de manera mas fácil juegos para los micro ordenadores Zx Spectrum, sin necesidad de saber BASIC o código maquina.

Puedes encontrar un tutorial de como usar la churrera y crear tus propios juegos en http://www.elmundodelspectrum.com/taller.php


CHANGELOG
r1 (0.1):
	* Primera versión del instalador
r2:
	* Añadido el programa SevenuP

r3 (0.2):
	* Añadido el programa Mappy, versión modifica para la Churrera
	* Añadidos las utilidades mapcnv, mapcnvbin, msc, reordenator y sprcnv de los Mojons Twins
	* Añadido el script churrera_installer.sh para crear o actualizar un proyecto de la Churrera.

r4 (0.3):
	* Ahora se muestra un menú para escoger los elementos a instalar
	* Ahora cada instalador esta en un función aparte para una mayor claridad del código
	* También comprueba si esta instalado lo necesario para instalar las aplicaciones

r5 (0.4):
	* Añadidos Beepola y BeepFX
	* Corregidas las urls en churrera_retriever por cambio de carpeta donde se encuentra el paquete con la Churrera
	* Ahora si se escoge la opción 0 y una o mas opciones se ejecuta la instalación de todo y se cierra sin ejecutar las demás opciones

r6: (0.5):
	* El proyecto pasa a llamarse ahora Herramientas Churreras
	* Añadido instalador para las utilidades del emulador FUSE
	* Ahora make.sh te dice cuanto espacio en memoria te queda (se necesita tener instalado fuse emu utils)
	* La Churrera Retriever ahora actualiza o descarga la Churrera 3.99.2
	* Otros cambios menores

r7: (0.6):
	* Actualizados los enlaces a los archivos debido a los cambios hechos por Dropbox.
	* Actualizada la fecha del Copyright.


make.sh
#!/bin/bash
NAME=
echo -e "\e[32mCompilando $NAME"
echo -e "Compilando script\e[0m"
cd ../script
wine msc $NAME.spt msc.h 24
cp *.h ../dev
echo -e "\e[32mRegenerando mapa\e[0m"
cd ../map
mapcnv mapa.map 8 3 15 10 15 packed
cp -f mapa.h ../dev
echo -e "\e[32mCompilando juego\e[0m"
cd ../dev
zcc +zx -vn $NAME.c -o $NAME.bin -lndos -lsplib2 -zorg=24200
echo -e "\e[32mCreando la cinta\e[0m"
MAYUS=echo $NAME | tr '[:lower:]' '[:upper:]'
bas2tap -a10 -s${MAYUS} loader.bas loader.tap
bin2tap -o screen.tap -a 16384 loading.bin
bin2tap -o main.tap -a 24200 $NAME.bin
cat loader.tap screen.tap main.tap > $NAME.tap

echo -e "\e[32mCompilación terminada.\e[0m"

# Si tienes instalado las herramientas del emulador fuse, descomenta las siguientes lineas (quitar las #) para
# saber de cuanto espacio libre dispones.
#echo -e "\e[32mCalculando espacio disponible\e[0m"
#TAMANO=`tzxlist main.tap | grep "Data length" | tail -1 | grep -o '[0-9]*'`
#TOTAL=`expr $TAMANO + 24200`
#ZXWEIGHT=`expr 60654 - $TOTAL`
#echo "$TAMANO + 24200 = $TOTAL ( $ZXWEIGHT )"

echo "\e[32mLimpiando\e[0m"
rm loader.tap
rm screen.tap
rm main.tap
rm $NAME.bin


churrera_retriever.sh
#!/bin/bash
# Churrera
# (c) 2013 - 2017 Alfonso Saavedra "Son Link"
# Script para automatizar la creación de un nuevo proyecto para la churrera o actualizar uno existente.
OIFS="$IFS"
IFS=$'\n'

function ayuda {
	echo "Churrera Retriever"
	echo '(c) 2013 - 2017 Alfonso Saavedra "Son Link"'
	echo ' '
	echo "Crea un nuevo proyecto para la Churrera o actualiza uno a la nueva versión"
	echo " "
	echo "Opciones:"
	echo "-h, --help		muestra esta ayuda"
	echo "-u, --update DIR	actualiza un proyecto existente en la ruta especificada"
	echo "-n, --new NAME DIR	crea un nuevo proyecto con el nombre y en la ruta especificada"
	exit 0
}

function new_proyect {
	cd /tmp
	wget -nc https://www.dropbox.com/s/ygi088fenkvynq2/Churrera3.99.2.tar.gz
	if [ ! -d $2 ]; then
		mkdir -p $2
	fi
	tar xzfv Churrera3.99.2.tar.gz
	cp Churrera3.99.2/* -r $2
	cd $2/dev
	mv churromain.c $1.c
	sed -i s/NAME=/NAME=${1}/ make.sh
	IFS="$OIFS"
	exit 0
}

function update_proyect {
	if [ ! -d "$1" ]; then
		echo -e "\e[31mEl directorio $1 no existe o no es un directorio.\e[0m"
	elif [ -d "$1-viejo" ]; then
		echo -e "\e[31mEl directorio $1-viejo ya existe. Por favor, bórrelo o muévalo a otro directorio.\e[0m"
	else
		PROYECTDIR="$1"
		echo -e "\e[32mActualizando el proyecto. Espere un momento ...\e[0m"
		cd /tmp
		wget -nc https://www.dropbox.com/s/ygi088fenkvynq2/Churrera3.99.2.tar.gz
		tar xzfv Churrera3.99.2.tar.gz
		cd Churrera3.99.1
		NAME=$(grep "NAME=" ${PROYECTDIR}/dev/make.sh | sed s/NAME=//)
		mv dev/churromain.c dev/$NAME.c
		mv script/churromain.spt script/$NAME.spt
		DIRS2COPY=('gfx' 'map' 'enems')
		for dir in  "${DIRS2COPY[@]}"; do
			cp -r ${PROYECTDIR}/${dir}/* ${dir}/
		done
		mv dev/config.h dev/confih.h.bak
		cp -f ${PROYECTDIR}/dev/* dev
		mv dev/config.h.bak confih.h
		mv ${PROYECTDIR} ${PROYECTDIR}-viejo
		cd /tmp
		mv Churrera3.99.2 $1
		echo -e "\e[32mActualización terminada.\e[0m"
		echo "Edite el archivo dev/config.h segun lo tiene en ${PROYECTDIR}-viejo/dev/config.h"
		echo "Una vez editado y comprobado que esta todo correcto puede borrar ${PROYECTDIR}-viejo si lo desea."
		IFS="$OIFS"
	fi
}

if [ $# -ge 1 ]; then
	case "$1" in
		-h|--help)
			ayuda
			;;
		-u|--update)
			update_proyect $2
			;;
		-n|--new)
			if [ $# -eq 3 ]; then
				new_proyect $2 $3
			else
				ayuda
			fi
			;;
		*)
			ayuda
			;;
	esac
else
	ayuda
fi


churrera-installer.sh
#!/bin/bash

# Churrera installer 0.1
# (c) 2013 - 2017 Alfonso Saavedra "Son Link"
# Churrera (c) 2013 The Mojon twins
# Script bash para instalar los componentes de la Churrera para crear juegos para las Zx Spectrum
# Script bajo licencia GPLv3

COMMANDS=('wget' 'unzip' 'gcc'  'make' 'automake')
ERRORS=0
USED=()

# Funciones para cada tarea. Es para tener un poco de orden en el código

function z88dk {
	echo -e "\e[1m\e[32mCompilando z88dk\e[0m"
	wget -nc https://www.dropbox.com/s/koyxyued91m4okh/z88dk-splib2-1.10.1.tar.gz
	tar xzf z88dk-splib2-1.10.1.tar.gz
	cd z88dk-splib2-1.10.1
	make
	echo -e "\e[1m\e[32mInstalando z88dk\e[0m"
	sudo make prefix=/usr install

	echo -e "\e[1mAñada las siguientes lineas en el fichero .bashrc de su carpeta HOME, /etc/profile o cree un fichero ejecutable en /etc/profile.d/:\e[0m"
	echo "export ZCCCFG=/usr/share/z88dk/lib/config/"
	echo "export Z80_OZFILES=/usr/share/z88dk/lib/"
	echo -e "Pulse una tecla para seguir\n"
	read -n 1 -s
}

function bin2 {
	echo -e "\e[1m\e[32mCompilando bin2tap\e[0m"
	wget -nc http://zeroteam.sk/files/bin2tap13.zip
	unzip -p bin2tap13.zip bin2tap/bin2tap.c > bin2tap.c
	gcc bin2tap.c -o bin2tap
	echo -e "\e[1m\e[32mInstalando bin2tap\e[0m"
	sudo cp bin2tap /usr/bin
}

function bas2 {
	echo -e "\e[1m\e[32mCompilando bas2tap\e[0m"
	wget -nc ftp://ftp.worldofspectrum.org/pub/sinclair/tools/generic/bas2tap26-generic.zip
	unzip bas2tap26-generic.zip
	gcc -Wall -O2 bas2tap.c -o bas2tap -lm ; strip bas2tap
	echo -e "\e[1m\e[32mInstalando bas2tap\e[0m"
	sudo cp bas2tap /usr/bin
}

function sevenup {
	echo -e "\e[1m\e[32mCompilando SevenuP\e[0m"
	wget -nc http://metalbrain.speccy.org/SevenuP-v1.21src_WIP.zip
	unzip SevenuP-v1.21src_WIP.zip
	cd SRC
	make -f makefile.unx CC=g++
	echo -e "\e[1m\e[32mInstalando SevenuP\e[0m"
	sudo cp SevenuP /usr/bin
}

function mojonas {
	echo -e "\e[1m\e[32mInstalando las utilidades mojonas\e[0m"
	wget -nc https://www.dropbox.com/s/n3wglv8sw5jdpen/utilidades-mojonas-linux.tar.gz
	tar xvfz utilidades-mojonas-linux.tar.gz
	sudo mkdir -p /opt/churrera/bin
	sudo cp utilidades-mojonas/* /opt/churrera/bin
	echo "Añada la ruta /opt/churrera/bin al PATH"
	read -n 1 -s
}

function map {
	echo -e "\e[1m\e[32mInstalando Mappy\e[0m"
	wget -nc http://www.mojontwins.com/churrera/mt-mappy.zip
	unzip mt-mappy.zip
	sudo cp -r Mappy /opt/churrera
	echo "#!/bin/bash" | sudo tee /usr/bin/mappy
	echo "cd /opt/churrera/Mappy && wine mapwin.exe" | sudo tee -a /usr/bin/mappy
	sudo chmod +x /usr/bin/mappy
}

function Beepola {
	echo -e "\e[1m\e[32mInstalando Beepola\e[0m"
	wget -nc http://freestuff.grok.co.uk/beepola/Beepola_v1.06.01.zip
	unzip Beepola_v1.06.01.zip
	sudo mkdir -p /opt/churrera/beepola
	sudo cp Beepola.exe /opt/churrera/beepola
	echo "#!/bin/bash" | sudo tee /usr/bin/beepola
	echo "wine /opt/churrera/beepola/Beepola.exe" | sudo tee -a /usr/bin/beepola
	sudo chmod +x /usr/bin/beepola
}

function BeepFX {
	echo -e "\e[1m\e[32mInstalando BeepFX\e[0m"
	wget -nc https://www.dropbox.com/s/l2rjjxqwz9d89ag/beepfx.zip
	sudo mkdir -p /opt/churrera/beepfx && sudo unzip beepfx.zip -d /opt/churrera/
	echo "#!/bin/bash" | sudo tee /usr/bin/beepfx
	echo "wine /opt/churrera/beepfx/BeepFX" | sudo tee -a /usr/bin/beepfx
	sudo chmod +x /usr/bin/beepfx
}

function fuseutils {
	search_lib=$(whereis -b libspectrum.h | cut -c8-)
	echo $search_lib
	if [ -z  "$search_lib" ]; then
		echo -e "\e[1m\e[32mCompilando libspectrum\e[0m"
		wget -nc http://sourceforge.net/projects/fuse-emulator/files/libspectrum/1.1.1/libspectrum-1.1.1.tar.gz
		tar xvfz libspectrum-1.1.1.tar.gz
		cd libspectrum-1.1.1
		./configure --prefix=/usr
		make
		echo -e "\e[1m\e[32mInstalando libspectrum\e[0m"
		sudo make install
	fi
	echo -e "\e[1m\e[32mCompilando fuse utils\e[0m"
	wget -nc http://sourceforge.net/projects/fuse-emulator/files/fuse-utils/1.1.1/fuse-utils-1.1.1.tar.gz
	tar xzfv fuse-utils-1.1.1.tar.gz
	cd fuse-utils-1.1.1
	ff=$( ffmpeg -version | grep "ffmpeg version" | cut -d " " -f 3 | sed  "s/\.//g")
	if [ $ff -gt 200 ]; then
		wget -nc https://raw.github.com/pld-linux/fuse-utils/master/ffmpeg_enum_codecid.patch
		patch -p1 < ffmpeg_enum_codecid.patch
	fi

	./configure --prefix=/usr
	make
	echo -e "\e[1m\e[32mInstalando fuse utils\e[0m"
	sudo make install
}
#Comprobamos si esta instalado lo necesario
for comm in  "${COMMANDS[@]}"; do
	command -v $comm >/dev/null 2>&1 || { echo >&2 "No se encuentra ${comm}."; ERRORS=$[$ERRORS+1];}
done

# Si no esta al menos unos de los programas se muestra un mensaje y salimos
if [ $ERRORS -gt 0 ]; then
	echo -e "\e[1m\e[31mUno o mas programas requeridos para la instalación no están instalados.\e[0m\nPor favor, instálelos y vuelva a intentar"; exit 1
fi

echo -e "\e[1m\e[32mChurrera installer 0.4\e[0m"
echo -e "Bienvenid@ al instalador de las utilidades para la \e[1mChurrera\e[0m de los Mojon Twins."
echo "A lo largo del proceso se instalaran los componentes necesarios para el uso."
echo -e "Por favor, lea el \e[31mINSTALL\e[0m para conocer los programas y librerías necesarias para la compilación y uso de los componentes."
echo -e "\e[31mAlgunos componentes pueden estar disponibles desde los repositorios de su distribución. Si es así se recomienda instalarlos a excepción de z88dk ya que es necesario compilarlo con una librería extra."
echo -e "\e[1m\e[31mALERTA: no ejecute este script con permisos de root por motivos de seguridad.\e[0m"
echo -e "Seleccione los componentes a instalar escribiendo el numero separado por espacios:\nEjemplo: 1 4\n"
echo -e "\e[1m1: z88dk		2: bin2tap"
echo -e "3: bas2tap		4: SevenuP"
echo -e "5: Utilidades mojonas	6: Mappy"
echo -e "7: Beepola		8: BeepFX"
echo -e "9: Fuse Emu Utils"
echo -e "0: todo"
read -a OPTIONS -p 'Escriba su selección: ' -t 120
echo -e "\e[0m"

if [ -z $OPTIONS ]; then
	exit
fi

if [ ! -d /tmp/churrera ]; then
	mkdir -p /tmp/churrera
fi
cd /tmp/churrera

if [[ " ${OPTIONS[*]} " == *" 0 "* ]]; then
	z88dk
	bin2
	bas2
	sevenup
	mojonas
	map
	Beepola
	BeepFX
	fuseutils
	exit
fi

for opt in  "${OPTIONS[@]}"; do
	if [[ " ${used[*]} " == *" $opt "* ]]; then
		continue
	fi
	case $opt in
		1)
			z88dk
			USED+=(1)
			;;
		2)
			bin2
			USED+=(2)
			;;
		3)
			bas2
			USED+=(3)
			;;
		4)
			sevenup
			USED+=(4)
			;;
		5)
			mojonas
			USED+=(5)
			;;
		6)
			map
			USED+=(6)
			;;
		7)
			Beepola
			USED+=(7)
			;;
		8)
			BeepFX
			;;
		9)
			fuseutils
			;;
	esac
done

echo -e "\e[1m\e[32mInstalación terminada.\e[0m"
echo "Gracias por usarlo y esperamos que disfrutes haciendo tus juegos."
echo "Pulse una tecla para terminar."
read -n 1 -s
rm -rf /tmp/churrera
